packages/utils/src/date/Locale.ts

```diff
@@ -0,0 +1,556 @@
+          break;
+        // no default
+      }
+    }
+
+    this._regExp = new RegExp(`^${str}$`, 'i');
+    return this._regExp;
+  }
+
+  private _monthNames: FormattedNames[] | null = null;
+  /**
+   * Get month names
+   */
+  private get monthNames (): FormattedNames[] {
+    if (this._monthNames) {
+      return this._monthNames;
+    }
+    this._monthNames = this.formatMonthNames();
+    return this._monthNames;
+  }
+
+  private _weekdayNames: FormattedNames[] | null = null;
+  /**
+   * Get weekday names
+   */
+  private get weekdayNames (): FormattedNames[] {
+    if (this._weekdayNames) {
+      return this._weekdayNames;
+    }
+    this._weekdayNames = this.formatWeekdayNames();
+    return this._weekdayNames;
+  }
+
+  private _periodNames: FormattedNames[] | null = null;
+  /**
+   * Get period names
+   */
+  private get periodNames (): FormattedNames[] {
+    if (this._periodNames) {
+      return this._periodNames;
+    }
+    this._periodNames = this.formatPeriodNames();
+    return this._periodNames;
+  }
+
+  private _eraNames: FormattedNames[] | null = null;
+  /**
+   * Get era names
+   */
+  private get eraNames (): FormattedNames[] {
+    if (this._eraNames) {
+      return this._eraNames;
+    }
+    this._eraNames = this.formatEraNames();
+    return this._eraNames;
+  }
+
+  private _parts: DateTimeFormatPart[] | null = null;
+  /**
+   * Get parts for formatter
+   */
+  private get parts (): DateTimeFormatPart[] {
+    if (this._parts) {
+      return this._parts;
+    }
+    this._parts = this.formatter.formatToParts(Date.UTC(2018, 0, 1));
+    return this._parts;
+  }
+
+  private _resolvedFormat: Format | null = null;
+  /**
+   * Get resolved ISO string format
+   */
+  public get isoFormat (): Format {
+    if (this._resolvedFormat) {
+      return this._resolvedFormat;
+    }
+
+    const options = this.options;
+
+    const day = options.day !== undefined || options.dateStyle !== undefined;
+    const month = day || options.month !== undefined;
+    const year = day || month || options.year !== undefined;
+    const millisecond = options.fractionalSecondDigits !== undefined;
+    const second = millisecond || options.second !== undefined || options.timeStyle === 'full' || options.timeStyle === 'long' || options.timeStyle === 'medium';
+    const minute = second || options.minute !== undefined || options.timeStyle === 'short';
+    const hour = minute || options.hour !== undefined;
+
+    if (hour && !year) {
+      // time only formats
+      this._resolvedFormat = millisecond ? TimeFormat.HHmmssSSS : second ? TimeFormat.HHmmss : TimeFormat.HHmm;
+    }
+    else if (year && !hour) {
+      // date only
+      this._resolvedFormat = day ? DateFormat.yyyyMMdd : month ? DateFormat.yyyyMM : DateFormat.yyyy;
+    }
+    else {
+      // datetime or ambiguous
+      this._resolvedFormat = millisecond ? DateTimeFormat.yyyMMddTHHmmssSSS : second ? DateTimeFormat.yyyMMddTHHmmss : DateTimeFormat.yyyMMddTHHmm;
```

Comment:
What if `minute === false`?

Reply:
minutes do not matter if seconds are provided. There is no valid ISO value format, which includes seconds, but does not include minutes/hours, similarly there is no valid format that include hours only. So essentially the result will always be one of:

- yyyy
- yyyyMM
- yyyyMMdd
- HHmm
- HHmmss
- HHmmss.SSS
- yyyyMMddTHHmm
- yyyyMMddTHHmmss
- yyyyMMddTHHmmss.SSS

However when string is parsed, we, the missing bits are populated from the reference date.

Reply:
I see.